src/app/clusters/user-label-server/tests/TestUserLabelCluster.cpp

```diff
@@ -0,0 +1,186 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <pw_unit_test/framework.h>
+
+#include <app/clusters/user-label-server/user-label-cluster.h>
+
+#include <app/clusters/testing/AttributeTesting.h>
+#include <app/server-cluster/AttributeListBuilder.h>
+#include <app/server-cluster/DefaultServerCluster.h>
+#include <app/server-cluster/testing/TestServerClusterContext.h>
+#include <clusters/UserLabel/Attributes.h>
+#include <clusters/UserLabel/Metadata.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::UserLabel;
+using namespace chip::app::Clusters::UserLabel::Attributes;
+
+namespace {
+
+struct TestUserLabelCluster : public ::testing::Test
+{
+    static void SetUpTestSuite() { ASSERT_EQ(Platform::MemoryInit(), CHIP_NO_ERROR); }
+    static void TearDownTestSuite() { Platform::MemoryShutdown(); }
+};
+
+constexpr size_t kUserLabelFixedClusterCount = 2;
+
+class UserLabelClusterTest
+{
+    UserLabelCluster userLabel;
+
+public:
+    template <typename... Args>
+    UserLabelClusterTest(Args &&... args) : userLabel(std::forward<Args>(args)...)
+    {}
+
+    template <typename F>
+    void Check(F check)
+    {
+        check(userLabel);
+    }
+};
+
+class ReadAttribute
+{
+    UserLabelCluster & mCluster;
+    DataModel::ActionReturnStatus mStatus;
+
+public:
+    ReadAttribute(UserLabelCluster & cluster) : mCluster(cluster), mStatus(CHIP_NO_ERROR) {}
+
+    DataModel::ActionReturnStatus GetStatus() const { return mStatus; }
+
+    void operator()(DataModel::ReadAttributeRequest & request)
+    {
+        Platform::ScopedMemoryBufferWithSize<uint8_t> buffer;
+        ASSERT_NE(buffer.Alloc(1024).Get(), nullptr);
+
+        AttributeReportIBs::Builder attributeReportIBsBuilder;
+        TLV::TLVWriter reportWriter;
+        reportWriter.Init(buffer.Get(), buffer.AllocatedSize());
+        ASSERT_EQ(attributeReportIBsBuilder.Init(&reportWriter), CHIP_NO_ERROR);
+
+        AttributeValueEncoder encoder(attributeReportIBsBuilder, Access::SubjectDescriptor{}, request.path, 0 /* dataVersion */);
+
+        mStatus = mCluster.ReadAttribute(request, encoder);
+    }
+};
+
+class WriteAttribute
+{
+    UserLabelCluster & mCluster;
+    DataModel::ActionReturnStatus mStatus;
+
+public:
+    WriteAttribute(UserLabelCluster & cluster) : mCluster(cluster), mStatus(CHIP_NO_ERROR) {}
+
+    DataModel::ActionReturnStatus GetStatus() const { return mStatus; }
+
+    void operator()(DataModel::WriteAttributeRequest & request)
+    {
+        Platform::ScopedMemoryBufferWithSize<uint8_t> buffer;
+        ASSERT_NE(buffer.Alloc(1024).Get(), nullptr);
```

Comment:
Should we use a constant for this alloc?

Generally we probably want MTU max size.

Extract to a constant and create a TODO to place it in some common place. I imagine this is reusable.

Reply:
I think ASSERT does a return ... so maybe in this case VerifyOrDie is better.

Is there a way to give the max size with a default parameter instead? Most tests will not need such large payloads and we want to be able to test on embedded.